# Choosing the location for the borehole

We were provided with oil samples in three regions: in each 10,000 fields, where the quality of oil and the volume of its reserves were measured. We need to build a machine learning model that will help determine the region where mining will bring the greatest profit. Let's analyze the possible profit and risks via *Bootstrap.*

Order:

- Deposits are being searched for in the selected region, the values of the signs are determined for each;
- Build a model and estimate the volume of stocks;
- Choose the deposits with the highest estimates of values. The number of deposits depends on the company's budget and the cost of developing one borehole;
- The profit is equal to the total profit of the selected deposits.

## Data preparation:

### Libraries import:

In [1]:
import pandas as pd

import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
import numpy as np
from numpy.random import RandomState

In [2]:
df_0 = pd.read_csv('/Users/a1/Desktop/data_folder/geo_data/geo_data_0.csv')
df_1 = pd.read_csv('/Users/a1/Desktop/data_folder/geo_data/geo_data_1.csv')
df_2 = pd.read_csv('/Users/a1/Desktop/data_folder/geo_data/geo_data_2.csv')

### Primary data analysis:

Check dataset df_0:

In [3]:
df_0.shape

(100000, 5)

In [4]:
df_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [5]:
df_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [6]:
df_0.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [7]:
df_0.duplicated().sum()

0

Check dataset df_1:

In [8]:
df_1.shape

(100000, 5)

In [9]:
df_1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [10]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [11]:
df_1.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [12]:
df_1.duplicated().sum()

0

Check dataset df_2:

In [13]:
df_2.shape

(100000, 5)

In [14]:
df_2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [15]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [16]:
df_2.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

In [17]:
df_2.duplicated().sum()

0

Primary data analysis showed:

* the absence of missed data;

* the absence of complete duplicates;

* accurate column names;

* with the exception of the id column - numerical data values, which is important for training the model;

* column id won't be useful for training - we can delete it from datasets.

Delete column id from each datasets:

In [18]:
df_0 = df_0.drop('id', axis=1)

In [19]:
df_0.head()

f0        f1        f2     product
0  0.705745 -0.497823  1.221170  105.280062
1  1.334711 -0.340164  4.365080   73.037750
2  1.022732  0.151990  1.419926   85.265647
3 -0.032172  0.139033  2.978566  168.620776
4  1.988431  0.155413  4.751769  154.036647

In [20]:
df_1 = df_1.drop('id', axis=1)

In [21]:
df_1.head()

f0         f1        f2     product
0 -15.001348  -8.276000 -0.005876    3.179103
1  14.272088  -3.475083  0.999183   26.953261
2   6.263187  -5.948386  5.001160  134.766305
3 -13.081196 -11.506057  4.999415  137.945408
4  12.702195  -8.147433  5.004363  134.766305

In [22]:
df_2 = df_2.drop('id', axis=1)

In [23]:
df_2.head()

f0        f1        f2     product
0 -1.146987  0.963328 -0.828965   27.758673
1  0.262778  0.269839 -2.530187   56.069697
2  0.194587  0.289035 -5.586433   62.871910
3  2.236060 -0.553760  0.930038  114.572842
4 -0.515993  1.716266  5.899011  149.600746

## Training and model validation

### We will prepare 2 samples from the available data for each dataset, namely: training and validation.

In [24]:
target_0 = df_0['product']
feature_0 = df_0.drop('product', axis=1)
features_train_0, features_valid_0, target_train_0, target_valid_0 = train_test_split(
    feature_0, target_0, test_size=0.25, random_state=12345)

In [25]:
target_1 = df_1['product']
feature_1 = df_1.drop('product', axis=1)
features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split(
    feature_1, target_1, test_size=0.25, random_state=12345)

In [26]:
target_2 = df_2['product']
feature_2 = df_2.drop('product', axis=1)
features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split(
    feature_2, target_2, test_size=0.25, random_state=12345)

### Model training:

Create the function for getting characteristics:

In [27]:
def get_metrics(target_valid, predictions_valid):
    print('RMSE:', np.sqrt(mean_squared_error(target_valid, predictions_valid)))
    print('Average inventory of predicted raw materials:', predictions_valid.mean())
    pass

#### Region "0":

In [28]:
model_0 = LinearRegression()
model_0.fit(features_train_0, target_train_0) 

LinearRegression()

In [29]:
predict_0 = model_0.predict(features_valid_0)
get_metrics(target_valid_0, predict_0)

RMSE: 37.5794217150813
Average inventory of predicted raw materials: 92.59256778438035


#### Region "1":

In [30]:
model_1 = LinearRegression()
model_1.fit(features_train_1, target_train_1) 

LinearRegression()

In [31]:
predict_1 = model_1.predict(features_valid_1)
get_metrics(target_valid_1, predict_1)

RMSE: 0.893099286775617
Average inventory of predicted raw materials: 68.72854689544602


#### Region "2":

In [32]:
model_2 = LinearRegression()
model_2.fit(features_train_2, target_train_2) 

LinearRegression()

In [33]:
predict_2 = model_2.predict(features_valid_2)
get_metrics(target_valid_2, predict_2)

RMSE: 40.02970873393434
Average inventory of predicted raw materials: 94.96504596800489


As a result of training a Linear regression model on data from three regions , the following results were obtained:

* the largest average volume of reserves is predicted in region "2": 94 thousand barrels, followed by region "0" : 92 thousand barrels and region "1" with a volume of 68 thousand barrels;

* at the same time, it is important to note that the model showed the best result on the data of the region "0", since its model showed the lowest RMSE = 37.6

## Preparation for profit calculation

We have the following info:
* During the exploration of the region, 500 points are explored, from which 200 best ones are selected for development using machine learning.

* The budget for the development of boreholes in the region is 10 billion.

* At current prices, one barrel of raw materials brings 450 units of income. The revenue from each unit of product is 450 thousand, since the volume is indicated in thousands of barrels.

Prepare the variables:

In [34]:
BUDGET = 10_000_000_000
PRICE_FOR_1000_BARREL = 450_000
POINTS=200

We need to determine a sufficient amount of raw materials for the break-even development of a new barehole:

In [35]:
MIN_PRODUCT = BUDGET / PRICE_FOR_1000_BARREL / POINTS
print('Sufficient volume of raw materials for break-even development of a new barehole', MIN_PRODUCT)

Sufficient volume of raw materials for break-even development of a new barehole 111.11111111111111


Sufficient volume of raw materials for break-even development of a new barehole, equal to 111 thousand barrels, exceeds the predicted volumes of bareholes in each region, the average forecasts of which are:

* region "0" - 92 thousand barrels;
* region "1" - 68 thousand barrels;
* region "2" - 94 thousand barrels.

## Calculation of profit and risks

After assessing the risks, we need to keep only those regions in which the probability of losses is less than 2.5%. Among them, the region with the highest average profit is chosen.

Create the function to calculate the profit:

In [36]:

def revenue(target, predict):
    
    predict_sorted = predict.sort_values(ascending=False)
    selected = target[predict_sorted.index][:200]
    return (selected.sum() * PRICE_FOR_1000_BARREL) - BUDGET


def bootstrap(target_valid, predict, region):
    state = np.random.RandomState(12345)
    values = []
    target = target_valid.reset_index(drop=True)
    predictions = pd.Series(predict)
    count = 0 
    for i in range(1000):
        target_subsample = target.sample(500, replace=True, random_state=state)
        predictions_subsample = predictions.loc[target_subsample.index.to_list()]
        target_subsample = target_subsample.reset_index(drop=True)
        predictions_subsample = predictions_subsample.reset_index(drop=True)
        value = revenue(target_subsample, predictions_subsample)
        values.append(value)
        
        if value < 0: 
            count += 1 
        
    values = pd.Series(values)
    risk = (values <0).mean()
    
    lower = values.quantile(0.025)
    upper = values.quantile(0.975) 
    
    print('Average profit on', region, 'region is:', round(values.mean(), 2), 'with a risk', 
          round(risk * 100, 2), '%')
    print('The confidence interval in the region is from', round(lower, 2),'to', round(upper, 2))

In [37]:
bootstrap(target_valid_0, predict_0, 0)

Average profit on 0 region is: 396164984.8 with a risk 6.9 %
The confidence interval in the region is from -111215545.89 to 909766941.55


In [38]:
bootstrap(target_valid_1, predict_1, 1)

Average profit on 1 region is: 456045105.79 with a risk 1.5 %
The confidence interval in the region is from 33820509.4 to 852289453.87


In [39]:
bootstrap(target_valid_2, predict_2, 2)

Average profit on 2 region is: 404403866.57 with a risk 7.6 %
The confidence interval in the region is from -163350413.4 to 950359574.92


Thus, thanks to the bootstrap technique, it was found that the "1" region is the safest for development, since its development is associated with the lowest risk of 1.5% and a confidence interval from 33820509.4 to 852289453.87

## Conclusion:

During the work on the project, data was prepared for work, models were trained on three different samples, and the work of these models was evaluated. Further, these results were additionally verified by the bootstrap technique. As a result, the region "1" turned out to be the safest from the point of view of profitability, since its development is associated with the lowest risk of 1% and a confidence interval from 68873225.37 to 931547591.26.